lib/puppet/functions/nomad/sorted_json.rb

```diff
@@ -0,0 +1,159 @@
+require 'json'
+
+Puppet::Functions.create_function(:'nomad::sorted_json') do
+  # This function takes unsorted hash and outputs JSON object making sure the keys are sorted.
+  # Optionally you can pass 2 additional parameters, pretty generate and indent length.
+  #
+  # *Examples:*
+  #
+  #     -------------------
+  #     -- UNSORTED HASH --
+  #     -------------------
+  #     unsorted_hash = {
+  #       'client_addr' => '127.0.0.1',
+  #       'bind_addr'   => '192.168.34.56',
+  #       'start_join'  => [
+  #         '192.168.34.60',
+  #         '192.168.34.61',
+  #         '192.168.34.62',
+  #       ],
+  #       'ports'       => {
+  #         'rpc'   => 8567,
+  #         'https' => 8500,
+  #         'http'  => -1,
+  #       },
+  #     }
+  #
+  #     -----------------
+  #     -- SORTED JSON --
+  #     -----------------
+  #
+  #     nomad::sorted_json(unsorted_hash)
+  #
+  #     {"bind_addr":"192.168.34.56","client_addr":"127.0.0.1",
+  #     "ports":{"http":-1,"https":8500,"rpc":8567},
+  #     "start_join":["192.168.34.60","192.168.34.61","192.168.34.62"]}
+  #
+  #     ------------------------
+  #     -- PRETTY SORTED JSON --
+  #     ------------------------
+  #     Params: data <hash>, pretty <true|false>, indent <int>.
+  #
+  #     nomad::sorted_json(unsorted_hash, true, 4)
+  #
+  #     {
+  #         "bind_addr": "192.168.34.56",
+  #         "client_addr": "127.0.0.1",
+  #         "ports": {
+  #             "http": -1,
+  #             "https": 8500,
+  #             "rpc": 8567
+  #         },
+  #         "start_join": [
+  #             "192.168.34.60",
+  #             "192.168.34.61",
+  #             "192.168.34.62"
+  #         ]
+  #     }
+  #
+  def sorted_json(unsorted_hash = {}, pretty = false, indent_len = 4)
+    quoted = false
+    # simplify jsonification of standard types
+    simple_generate = lambda do |obj|
```

Comment:
In the 4.x/modern function API there is no need to do lambdas like this - the logic can be in a separate method - i.e. `def simple_generate(obj)` that the main method called by the function's dispatcher can call.
having separate methods makes it a lot easier to read the code and provides a more natural place to put documentation of the helper logic.

Reply:
@hlindberg can you point me to some examples please, as to how i could go about implementing this in a more modern way?

Reply:
it is easy instead of:
```
simple_generate = lambda do | obj | { ... }
sorted_generate = lambda do | obj | { ... }
sorted_pretty_generate = lambda do | obj | { }

// ...
    if pretty
      return sorted_pretty_generate.call(unsorted_hash, indent_len) << "\n"
    else
      return sorted_generate.call(unsorted_hash)
    end
```

You would do:
```
def simple_generate(obj)
  # body of this existing lambda
end
def sorted_generate(obj)
  # body of this existing lambda
end
def sorted_pretty_generate(obj)
  # body of this existing lambda
end
```
(Also needs to change the calls to the `simple_generate` to be just a simple method call, i.e. `simple_generat(obj)`
Thus making the method that gets called be only this part:
```
def sorted_json(unsorted_hash = {}, pretty = false, indent_len = 4)
  quoted = false
    if pretty
      return sorted_pretty_generate(unsorted_hash, indent_len) << "\n"
    else
      return sorted_generate(unsorted_hash)
    end
```
 This because the 4.x API allows you do define as many methods as you like inside of the function whereas the 3.x API forced you to jump through hoops simulating that by using lambdas.

Hope this helps.